### ISSUE TITLE
Right and Middle mouse button reversed

### ISSUE THREAD

**Issue by baryluk** [univrsal/input-overlay#180]
**Describe the bug**

When clicking right mouse button (button 3), middle one is highlighted.

When clicking middle mouse button (button 2), right one is highlighted.

Left (button 1), and side buttons (buttons 8, 9), and scroll wheel (buttons 4 and 5) works

**To Reproduce**

input-overlay.v5.0.0rc1-8d04a98.linux.zip

input-overlay.v5.0.0rc1-8d04a98.presets.zip

Add InputOverlay, then select a `mouse-no-movement.json` as layout.

**Expected behaviour**

All buttons should highlight correctly

**Screenshots**

Me pressing right mouse button:

![Screenshot at 2021-06-06 17-43-53](https://user-images.githubusercontent.com/110322/120934701-f5020b00-c6ee-11eb-9114-b9e114c8b5d6.png)

**Additional information:**
- Plugin version: v5.0.0rc1
- OBS 26.1.2+dfsg1-1+b2 (linux), amd64
- Full OBS Studio log: https://obsproject.com/logs/T-kCdqB0HvviUlNW

**Additional context**

- Linux, Debian, testing, OBS 26.1.2+dfsg1-1+b2, amd64
- Linux 5.10.28
- X11 (xorg-core 1.20.11)
- Mate (marco) 1.24.1
- libinput10:amd64 1.16.4-3
- xserver-xorg-input-libinput 0.30.0-1
- libxcb-xinput 1.14-3
- xinput 1.6.3-1
- libevdev2 1.11.0+dfsg-1

```
$ lsusb | grep Corsair
Bus 001 Device 010: ID 1b1c:1b30 Corsair Corsair Vengeance M60 Mouse
```

```
$ xinput get-button-map 14
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 
$ xinput get-button-map 16
1 2 3 4 5 6 7 
$
```

```

**Comment by kpuljek** [comment]
Same here, the mouse works as expected in the web editor, but in OBS the middle mouse and the right click are reversed.

**Comment by berglh** [comment]
I am experience the same problem as the original post on 5.0 pre-release, using a texture I made for the G Pro Wireless. Working fine in the web editor (cct), in OBS I'm setting right click as middle mouse button. Here are my configs for testing:

**Texture**:
![G Pro Wireless Texture](https://user-images.githubusercontent.com/12455338/140633459-03806469-f5f8-43af-ae69-86f39af1b9c9.png)

**Config**:
```json
{
    "default_width": 0,
    "default_height": 0,
    "space_h": 0,
    "space_v": 0,
    "flags": 8,
    "overlay_width": 209,
    "overlay_height": 370,
    "elements": [
        {
            "type": 3,
            "pos": [
                32,
                8
            ],
            "id": "mouse button0",
            "z_level": 0,
            "mapping": [
                1,
                1,
                80,
                161
            ],
            "code": 1
        },
        {
            "type": 3,
            "pos": [
                112,
                8
            ],
            "id": "mouse button1",
            "z_level": 0,
            "mapping": [
                81,
                1,
                81,
                161
            ],
            "code": 2
        },
        {
            "type": 3,
            "pos": [
                15,
                121
            ],
            "id": "mouse button2",
            "z_level": 0,
            "mapping": [
                162,
                1,
                13,
                60
            ],
            "code": 5
        },
        {
            "type": 3,
            "pos": [
                20,
                175
            ],
            "id": "mouse button3",
            "z_level": 0,
            "mapping": [
                163,
                125,
                12,
                50
            ],
            "code": 4
        },
        {
            "type": 4,
            "pos": [
                100,
                50
            ],
            "id": "mouse wheel4",
            "z_level": 0,
            "mapping": [
                164,
                231,
                24,
                54
            ]
        },
        {
            "type": 3,
            "pos": [
                196,
                121
            ],
            "id": "mouse button5",
            "z_level": 0,
            "mapping": [
                175,
                1,
                13,
                60
            ],
            "code": 7
        },
        {
            "type": 3,
            "pos": [
                193,
                175
            ],
            "id": "mouse button6",
            "z_level": 0,
            "mapping": [
                176,
                125,
                12,
                50
            ],
            "code": 6
        },
        {
            "type": 0,
            "pos": [
                18,
                47
            ],
            "id": "texture7",
            "z_level": 0,
            "mapping": [
                270,
                1,
                188,
                323
            ]
        }
    ]
}
```

**Comment by berglh** [comment]
@univrsal Hi Alex,

I think I've figured out something about what is happening here. So the mouse wheel (type 4) seem to assume that the middle mouse button code is code 3. When I change the right mouse button to code below, it shows both the right mouse button and the middle mouse button being depressed in the display when I click the right mouse button. My guess is that the codes actually being sent by the mouse for mouse two is actually code 3 and the middle mouse button is code 2 in the libinput sides of things.

https://user-images.githubusercontent.com/12455338/141778986-784cb221-d788-4789-91d9-cf820212d3fa.mp4

```diff
        {
            "type": 3,
            "pos": [
                112,
                8
            ],
            "id": "mouse button1",
            "z_level": 0,
            "mapping": [
                81,
                1,
                81,
                161
            ],
-           "code": 2
+           "code": 3
        },
```
So, I ran:

```bash
$ xev | grep ', button'
    state 0x10, button 1, same_screen YES               # left mouse button
    state 0x110, button 1, same_screen YES
    state 0x10, button 3, same_screen YES               # right mouse button
    state 0x410, button 3, same_screen YES
    state 0x10, button 2, same_screen YES               # mouse wheel button
    state 0x210, button 2, same_screen YES
    state 0x10, button 4, same_screen YES               # scroll-up wheel
    state 0x810, button 4, same_screen YES
    state 0x10, button 5, same_screen YES               # scroll-down wheel
    state 0x1010, button 5, same_screen YES
    state 0x10, button 9, same_screen YES               # left side mouse forward
    state 0x10, button 9, same_screen YES
    state 0x10, button 8, same_screen YES               # left side mouse back
    state 0x10, button 8, same_screen YES
    state 0x10, button 10, same_screen YES             # right side mouse forward
    state 0x10, button 10, same_screen YES
    state 0x10, button 11, same_screen YES             # right side mouse back
    state 0x10, button 11, same_screen YES
```

This is strange, because I set the codes in the web config program, and it definitely saw right mouse button as code 2. Anyway, not sure if this will help with debugging the problem here, or if it's a libinput issue in GNU/Linux in general.

A quick work around would be allowing the setting of the "mouse wheel" mouse button code in the config file.
Long term, would be figuring out why the mouse codes in Chrome don't seem to match the codes in `xev`.
i.e:
```diff
        {
            "type": 4,
            "pos": [
                100,
                50
            ],
            "id": "mouse wheel4",
            "z_level": 0,
            "mapping": [
                164,
                231,
                24,
                54
-            ]
+          ],
+          "code": 2
        },
```

**Comment by jens-br** [comment]
Hi all,
same here (Linux, OBS 27.1.3, input-overlay 5.0.0rc1). Do you have any updates on this? Could you include this fix in the upcoming version 5.0 of input-overlay?
We can currently only use a workaround which uses just the left and the right mouse buttons, but ignores the mouse wheel completely. It is shown in the online editor correctly, but not in OBS. It would be very helpful for our project if the mouse wheel would be working as well.

Thank you very much for your work! :+1: 
Merry Christmas and best regards! :-)

**Comment by 3laws** [comment]
Is this still being worked on? I can't manage to make the workaround work for me.

**Comment by univrsal** [comment]
It should already be fixed as this was mostly an issue with libuiohook if I recall correctly, there's just no release for it. Might wanna try the latest nightly https://github.com/univrsal/input-overlay/actions/runs/2384290822

**Comment by Naushikha** [comment]
Hi guys!
This seems like an old thread but if you are like me stuck on Linux Mint 20 (or similar distro) with no way to get the new updated plugin to work (because it requires Qt6 [#283], and Linux Mint only supports Qt5 currently), 
I got a workaround that would fix the mouse with scroll (without displaying middle-click).

[mouse_partially_fixed.zip](https://github.com/univrsal/input-overlay/files/9890285/mouse_partially_fixed.zip)

Basically I edited the json config to make it work for the right click and removed the middle-click portion from the image,
![image](https://user-images.githubusercontent.com/56594989/198687914-61ef521b-333a-4265-98cb-52a79a2ca79f.png)

It works without the middle-click and for my use-case it was better than having to install and configure Qt6 on my system.
Hope this helps someone!
